### ISSUE TITLE
Add Equalized Odds for subgroup fairness analysis

### ISSUE THREAD

**Issue by Khanz9664** [Khanz9664/TrustLens#17]
### 🎯 Objective
Add Equalized Odds to better evaluate fairness across subgroups by measuring differences in True Positive Rate (TPR) and False Positive Rate (FPR).

---

### 🧐 Why This Matters

Accuracy alone can hide important fairness issues.

A model might have similar accuracy across groups but still behave very differently:
- one group gets more false negatives (low TPR)
- another gets more false positives (high FPR)

Equalized Odds helps surface these differences.

---

### 🛠️ Task

Implement:

```python
equalized_odds(y_true, y_pred, sensitive_features) -> dict
```

in:
```python
trustlens/metrics/bias.py
```

### 📦 Expected Output
```python
result = equalized_odds(y_true, y_pred, {"gender": gender_arr})

# {
#   "gender": {
#     "M": {"tpr": 0.88, "fpr": 0.12},
#     "F": {"tpr": 0.61, "fpr": 0.08},
#     "__summary__": {
#       "tpr_gap": 0.27,
#       "fpr_gap": 0.04,
#       "violation": "severe"
#     }
#   }
# }
```

### ⚖️ Violation Thresholds
`> 0.15` → "severe"
`0.05–0.15` → "moderate"
`< 0.05` → "acceptable"

### 💡 Implementation Hints
- Use sklearn.metrics:
    - recall_score for TPR
    - confusion_matrix or manual computation for FPR

Follow existing patterns in:
```python
trustlens/metrics/bias.py
```
- Ensure output is JSON-serializable
- Handle edge cases (e.g., division by zero)

### 📚 Reference
Hardt, Price, Srebro (2016)
Equality of Opportunity in Supervised Learning (NeurIPS)

### ✅ Acceptance Criteria
 - [x] Function implemented with correct API
 - [x] Per-group TPR and FPR computed
 - [x] Summary metrics (gaps + violation level) included
 - [x] Unit tests added (test_bias.py)
 - [x] Edge cases handled
 - [x] Documentation updated

### 🙌 Contribution

If you’d like to work on this, please comment before starting so we can coordinate.
Happy to guide implementation if needed 🚀

**Comment by komoike-oss28-ui** [comment]
Hi, I'd like to work on this!

I've reviewed `bias.py` and the existing patterns (e.g. `subgroup_performance`). My plan:

1. Implement `equalized_odds(y_true, y_pred, sensitive_features)` following the same structure as `subgroup_performance` — iterating over features and groups via `np.unique`
2. Compute per-group TPR with `recall_score` and FPR manually via `confusion_matrix`, consistent with the existing sklearn usage
3. Add `__summary__` with `tpr_gap`, `fpr_gap`, and violation level (`severe` / `moderate` / `acceptable`)
4. Handle edge cases (zero division, single-group input) similarly to the `zero_division=0` pattern already in use
5. Add unit tests in `test_bias.py` and update docs

Happy to open a draft PR early for feedback. Is this still available?

**Comment by Khanz9664** [comment]
Hey @komoike-oss28-ui — this is a great breakdown, thanks for taking the time to understand the codebase 👍

Your plan looks solid and aligns well with the existing design. Happy to assign this to you 

A few small suggestions to make the implementation consistent with TrustLens:

---

###  Implementation Notes

**1. Structure (important)**
Yes—please follow the same pattern as `subgroup_performance`, including:

* iterating over `sensitive_features`
* grouping via `np.unique`
* returning a nested dict per feature

---

**2. TPR / FPR computation**

* TPR → `recall_score(y_true, y_pred, zero_division=0)` (per group)
* FPR → recommend computing manually from confusion matrix:

FPR = FP/(FP + TN)

Make sure to guard against division by zero (return `0.0` if denominator is 0).

---

**3. Summary Block**
For `__summary__`:

* `tpr_gap = max(tpr) - min(tpr)`
* `fpr_gap = max(fpr) - min(fpr)`

Violation levels:

* `> 0.15` → `"severe"`
* `0.05–0.15` → `"moderate"`
* `< 0.05` → `"acceptable"`

---

**4. Edge Cases**
Please handle:

* single subgroup → gaps should be `0.0`, violation `"acceptable"`
* empty slices / zero positives / zero negatives
* ensure everything is JSON-serializable (plain floats, no numpy types)

---

**5. Tests**
In `test_bias.py`, include:

* normal multi-group case
* single-group case
* edge case (no positives or no negatives)
* check both per-group values and summary outputs

---

**6. Keep it lightweight**
Avoid adding new dependencies—stick to `numpy` + `sklearn` only.

---

### 💡 Optional (nice-to-have)

If easy, keep the function extensible for future fairness metrics (same input/output pattern).

---

Feel free to open a **draft PR early** if you want feedback on structure before finalizing.

Looking forward to this—this is an important feature for TrustLens 🔥
